# Nama : Adriati Manuk Allo
# NIM  : 2209106018
# Kelas: A1 22

## Posttest 4 
## Praktikum Mata Kuliah Kecerdasan Buatan

## Data Cleaning

#### SOAL : Lakukan Data cleaning pada data dengan nilai null (jika tidak ada nilai null pada dataset, maka buatlah menjadi ada). Ganti nilai null sesuai ketentuan. (bilangan bulat : median/modus, bilangan desimal : mean, tulisan : modus). Nilai plus jika strategy yang digunakan menggunakan 3 nilai

In [1]:
from sklearn.impute import SimpleImputer
import pandas as pd

In [2]:
df = pd.read_csv('ford.csv')
df.head()

model  year  price transmission  mileage fuelType  tax   mpg  engineSize
0   Fiesta  2017  12000    Automatic    15944   Petrol  150  57.7         1.0
1    Focus  2018  14000       Manual     9083   Petrol  150  57.7         1.0
2    Focus  2017  13000       Manual    12456   Petrol  150  57.7         1.0
3   Fiesta  2019  17500       Manual    10460   Petrol  145  40.3         1.5
4   Fiesta  2019  16500    Automatic     1482   Petrol  145  48.7         1.0

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17966 entries, 0 to 17965
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         17966 non-null  object 
 1   year          17966 non-null  int64  
 2   price         17966 non-null  int64  
 3   transmission  17966 non-null  object 
 4   mileage       17966 non-null  int64  
 5   fuelType      17966 non-null  object 
 6   tax           17966 non-null  int64  
 7   mpg           17966 non-null  float64
 8   engineSize    17966 non-null  float64
dtypes: float64(2), int64(4), object(3)
memory usage: 1.2+ MB


In [4]:
df.isnull().sum()

model           0
year            0
price           0
transmission    0
mileage         0
fuelType        0
tax             0
mpg             0
engineSize      0
dtype: int64

Karena setelah di cek tidak ada nilai null, maka kita akan membuat data menjadi ada nilai nullnya

Untuk membuat datanya menjadi null, buka file "ford.csv" di excel kemudian hapus beberapa record (kolomnya terserah)

Membaca kembali file yang datanya sudah ada nilai nullnya

In [45]:
df = pd.read_csv('ford.csv')
df.head()

model  year  price transmission  mileage fuelType    tax   mpg  \
0   Fiesta   217  12000    Automatic    15944   Petrol  150.0  57.7   
1    Focus  2018  14000       Manual     9083   Petrol  150.0  57.7   
2    Focus  2017  13000       Manual    12456   Petrol  150.0  57.7   
3   Fiesta  2019  17500       Manual    10460   Petrol  145.0  40.3   
4   Fiesta  2019  16500    Automatic     1482   Petrol  145.0  48.7   

   engineSize  
0         1.0  
1         1.0  
2         1.0  
3         1.5  
4         1.0

In [46]:
df.isnull().sum()

model            0
year             0
price            0
transmission     0
mileage          0
fuelType         8
tax             18
mpg              0
engineSize      10
dtype: int64

Dapat dilihat, sekarang kolom tax terdapat 18 record yang nilainya null, kolom engineZine terdapat 10 record yang nilainya null, dan kolom fuelType terdapat 8 record yang nilainya null.

Mengganti tax dengan mean

In [47]:
si_mean = SimpleImputer(strategy='mean')
df['tax'] = si_mean.fit_transform(df[['tax']])

In [48]:
df['tax'].mean()

113.42695564965456

In [49]:
df.isnull().sum()

model            0
year             0
price            0
transmission     0
mileage          0
fuelType         8
tax              0
mpg              0
engineSize      10
dtype: int64

Nilai null pada kolom tax berhasil diganti dengan nilai mean

In [50]:
df[['fuelType']].describe()

fuelType
count     17958
unique        5
top      Petrol
freq      12173

Mengganti fuelType dengan nilai modus

In [51]:
si_most_frequent = SimpleImputer(strategy='most_frequent')
df[['fuelType']] = si_most_frequent.fit_transform(df[['fuelType']])

In [52]:
df.isnull().sum()

model            0
year             0
price            0
transmission     0
mileage          0
fuelType         0
tax              0
mpg              0
engineSize      10
dtype: int64

Nilai null pada kolom fuelType berhasil diganti dengan nilai modus

Mengganti engineSize dengan nilai median

In [53]:
si_median = SimpleImputer(strategy='median')
df[['engineSize']] = si_median.fit_transform(df[['engineSize']])

In [54]:
df.isnull().sum()

model           0
year            0
price           0
transmission    0
mileage         0
fuelType        0
tax             0
mpg             0
engineSize      0
dtype: int64

Nilai null pada kolom engineSize berhasil diganti dengan nilai median

#### SOAL : Lakukan Data cleaning pada data dengan nilai duplikat. (Jika tidak ada nilai duplikat pada dataset, maka buatlah menjadi ada)

In [55]:
df. duplicated().sum()

152

Dapat dilihat bahwa jumlah data duplikat pada dataset "ford" sebanyak 152 data

Untuk menghapus nilai duplikat kita dapat menggunakan command "drop_duplicates" dari pandas

In [56]:
df.drop_duplicates(inplace=True)

In [57]:
df. duplicated().sum()

0

Setelah di hapus dan dilakukan pengecekan kembali dengan menggunakan command "df_duplicated().sum()" jumlah data duplikat menjadi 0

## Data Transformation

#### SOAL : Lakukan normalisasi data pada salah satu attribute menggunakan Min Max scaler (buatlah copy dataset terlebih dahulu)

##### Normalisasi

In [59]:
from sklearn.preprocessing import MinMaxScaler

In [90]:
df_mm = df.copy()
df_mm.dtypes

model            object
year              int64
price             int64
transmission     object
mileage           int64
fuelType         object
tax             float64
mpg             float64
engineSize      float64
dtype: object

In [92]:
mm = MinMaxScaler()
df_mm[['price']] = mm.fit_transform(df_mm[['price']])

print("Kolom price sebelum normalisasi:")
print("Max: ", df['price'].max())
print("Min: ",df['price'].min(), "\n")

print("Kolom price setelah normalisasi:")
print("Max: ", df_mm['price'].max())
print("Min: ",df_mm['price'].min(), "\n")

Kolom price sebelum normalisasi:
Max:  54995
Min:  495 

Kolom price setelah normalisasi:
Max:  1.0
Min:  0.0 



#### SOAL : Lakukan standarisasi pada dataset (buatlah copy dataset terlebih dahulu)

##### Standarisasi

In [100]:
from sklearn.preprocessing import StandardScaler

In [101]:
df_ss = df.copy()
df_ss.dtypes

model            object
year              int64
price             int64
transmission     object
mileage           int64
fuelType         object
tax             float64
mpg             float64
engineSize      float64
dtype: object

In [102]:
ss =  StandardScaler()
df_ss[['price']] = mm.fit_transform(df_ss[['price']])

print("Kolom price sebelum standarisasi:")
print("Mean: ", df['price'].mean())
print("Standar Deviasi: ",df['price'].std(), "\n")

print("Kolom price setelah standarisasi:")
print("Mean: ", df_ss['price'].mean())
print("Standar Deviasi: ",df_ss['price'].std(), "\n")

Kolom price sebelum standarisasi:
Mean:  12269.261872684405
Standar Deviasi:  4736.122576097706 

Kolom price setelah standarisasi:
Mean:  0.21604150225109
Standar Deviasi:  0.08690133167151755 



#### SOAL : Ganti tipe data salah satu attribute angka

In [107]:
df['mileage'] = df['mileage'].astype('float')

In [108]:
df.dtypes

model            object
year              int64
price             int64
transmission     object
mileage         float64
fuelType         object
tax             float64
mpg             float64
engineSize      float64
dtype: object

## Encoding

#### SOAL : Lakukan one hot encoding pada dataset yang kalian miliki

In [109]:
from sklearn.preprocessing import OneHotEncoder

In [116]:
pd.get_dummies(df['fuelType'].replace(False,0).replace(True,1))


Diesel  Electric  Hybrid  Other  Petrol
0       False     False   False  False    True
1       False     False   False  False    True
2       False     False   False  False    True
3       False     False   False  False    True
4       False     False   False  False    True
...       ...       ...     ...    ...     ...
17961   False     False   False  False    True
17962   False     False   False  False    True
17963    True     False   False  False   False
17964   False     False   False  False    True
17965   False     False   False  False    True

[17814 rows x 5 columns]